### PR TITLE
Add new attribute RaiseWarn

### PR DESCRIPTION
**Pull request by pali**
The RaiseWarn attribute can be used to force warnings to raise exceptions
rather then simply printing them. It is off by default.

Similar to attribute RaiseError.

### CODE REVIEW

File: DBI.pm [hunk 1]
```diff
@@ -3789,6 +3790,24 @@ By default, C<DBI-E<gt>connect> sets C<PrintError> "on".
...
+then simply printing them. It is "off" by default.
+When set "on", any method which sets warning condition will cause
+the DBI to effectively do a C<die("$class $method warning: $DBI::errstr")>,
+where C<$class> is the driver class and C<$method> is the name of the method
+that sets warning condition. E.g.,
+
+  DBD::Oracle::db execute warning: ... warning text here ...
+
+If you turn C<RaiseWarn> on then you'd normally turn C<PrintWarn> off.
+If C<PrintWarn> is also on, then the C<PrintWarb> is done first (naturally).
```
Comment (timbunce) [inline DBI.pm]:
```suggestion
If C<PrintWarn> is also on, then the C<PrintWarn> is done first (naturally).
```

**Comment by pali** [comment]
@timbunce I sent email to dbi-dev mailing list.

There is question about changing behavior or Print/RaiseError, so this is something for you to resolve.

File: DBI.xs [hunk 2]
```diff
@@ -4066,7 +4073,7 @@ XS(XS_DBI_dispatch)
         }
 
         hook_svp = NULL;
-        if (    SvTRUE(err_sv)
+        if (   (SvTRUE(err_sv) || (is_warning && DBIc_has(imp_xxh, DBIcf_RaiseWarn)))
```
Comment (timbunce) [inline DBI.xs]: With this change HandleError will be called for warnings but only if RaiseWarn is true.
That seems surprising, and also not mentioned in the docs.
You could revert this hunk and remove the mention of !hook_svp in the hunk below, or update the docs and provide a good rationale for the behaviour.
I think I'd prefer the former.

Comment (pali) [reply DBI.xs]: Currently `HandleError` is **not** called for warnings even when `PrintWarn` is set to true.

My change adds a new `RaiseWarn` attribute which is by default false. So default behavior of `HandleError` is not changed, it is not called for warnings.

When `RaiseWarn` is set to true value, then all warnings are `fatal` similar to errors and I think it make sense to send them into `HandleError` callback. So users can handle in `HandleError` anything which is going to throw an exception from DBI. I think this is better behavior and expected behavior. If you agree I update information about it in documentation.

Comment (pali) [reply DBI.xs]: I updated documentation for `HandleError` and rebased this patch on top of master branch.

Comment (pali) [reply DBI.xs]: @timbunce Is it now better?

**Comment by pali** [comment]
Done! I created a new test file which test all combinations of `RaiseWarn`, `RaiseError` and `HandleError`.

**Comment by timbunce** [comment]
Thanks @pali!

**Pull request merged**